### PR TITLE
🌱 Implement KCP and MS Deleting conditions

### PR DESCRIPTION
**Pull request by sbueringer**
**What this PR does / why we need it**:
Needs a rebase after #11364 is merged

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of https://github.com/kubernetes-sigs/cluster-api/issues/11105

### CODE REVIEW

File: controlplane/kubeadm/api/v1beta1/v1beta2_condition_consts.go [hunk 1]
```diff
@@ -213,12 +213,27 @@ const (
...
-// KubeadmControlPlane's Paused condition and corresponding reasons that will be used in v1Beta2 API version.
-const (
-	// KubeadmControlPlanePausedV1Beta2Condition is true if this resource or the Cluster it belongs to are paused.
-	KubeadmControlPlanePausedV1Beta2Condition = clusterv1.PausedV1Beta2Condition
+	// KubeadmControlPlaneDeletingDeletionTimestampNotSetV1Beta2Reason surfaces when the KCP is not deleting because the
+	// DeletionTimestamp is not set.
+	KubeadmControlPlaneDeletingDeletionTimestampNotSetV1Beta2Reason = clusterv1.DeletionTimestampNotSetV1Beta2Reason
+
+	// KubeadmControlPlaneDeletingWaitingForWorkersDeletionV1Beta2Reason surfaces when the KCP deletion
+	// waits for the workers to be deleted.
```
Comment (fabriziopandini) [inline controlplane/kubeadm/api/v1beta1/v1beta2_condition_consts.go:222]: I'm not sure we should phrase this as "waiting for" or as "kcp cannot be deleted (or kcp deletion cannot start) when there are still workers"

File: controlplane/kubeadm/internal/controllers/controller.go [hunk 2]
```diff
@@ -587,15 +591,27 @@ func (r *KubeadmControlPlaneReconciler) reconcileClusterCertificates(ctx context
 	return nil
 }
 
+// scope holds the different objects that are read and used during the reconcile.
+type scope struct {
```
Comment (fabriziopandini) [inline controlplane/kubeadm/internal/controllers/controller.go:595]: What I'm doing in other PRs is to use controlPlane as a scope (and if fact this is what it is).
Let's keep doing this without introducing a new scope.

File: controlplane/kubeadm/internal/controllers/status.go [hunk 3]
```diff
@@ -167,7 +167,7 @@ func (r *KubeadmControlPlaneReconciler) updateV1beta2Status(ctx context.Context,
 	setMachinesReadyCondition(ctx, controlPlane.KCP, controlPlane.Machines)
 	setMachinesUpToDateCondition(ctx, controlPlane.KCP, controlPlane.Machines)
 	setRemediatingCondition(ctx, controlPlane.KCP, controlPlane.MachinesToBeRemediatedByKCP(), controlPlane.UnhealthyMachines())
-
+	setDeletingCondition(ctx, controlPlane.KCP, s.deletingReason, s.deletingMessage)
```
Comment (fabriziopandini) [inline controlplane/kubeadm/internal/controllers/status.go:170]: Let's drop deleting from the todo in L171

**Review by fabriziopandini — COMMENTED** [review]
first pass, I will make another one once it is rebased

File: controlplane/kubeadm/internal/controllers/controller.go (hunk 2 above)
Comment (sbueringer) [reply controlplane/kubeadm/internal/controllers/controller.go:595]: Okay, fine for me. Adding deletingReason & deletingMessage just seemed pretty strange to me. Considering what we store within ControlPlane today

File: controlplane/kubeadm/api/v1beta1/v1beta2_condition_consts.go (hunk 1 above)
Comment (sbueringer) [reply controlplane/kubeadm/api/v1beta1/v1beta2_condition_consts.go:222]: What I liked about this is the consistency with the Cluster controller

Comment (sbueringer) [reply controlplane/kubeadm/api/v1beta1/v1beta2_condition_consts.go:222]: What we could do is make this a bit clearer through the message, wdyt?

File: controlplane/kubeadm/internal/controllers/controller.go (hunk 2 above)
Comment (sbueringer) [reply controlplane/kubeadm/internal/controllers/controller.go]: Done

File: controlplane/kubeadm/api/v1beta1/v1beta2_condition_consts.go (hunk 1 above)
Comment (fabriziopandini) [reply controlplane/kubeadm/api/v1beta1/v1beta2_condition_consts.go:222]: ack, let's change the message

Comment (sbueringer) [reply controlplane/kubeadm/api/v1beta1/v1beta2_condition_consts.go:222]: Done

**Review by fabriziopandini — COMMENTED** [review]
Nice work! two more down in the todo list!
/lgtm
/approve

**Comment by sbueringer** [comment]
/test pull-cluster-api-e2e-main
If I see correctly two known MP flakes

**Pull request merged**